CMakeLists.txt

```diff
@@ -8,65 +8,74 @@
 #  directory.
 #
 
-cmake_minimum_required(VERSION 3.0.2)
+cmake_minimum_required(VERSION 3.10.2)
 
 project(Fenix C)
 # The version number.
 set(FENIX_VERSION_MAJOR 1)
 set(FENIX_VERSION_MINOR 0)
 
-option(BUILD_EXAMPLES  "Builds example programs from the examples directory"   OFF)
-option(BUILD_TESTING   "Builds tests and test modes of files"                  ON)
+option(FENIX_EXAMPLES    "Builds example programs from the examples directory" OFF)
+option(FENIX_TESTS       "Builds tests and test modes of files"                OFF)
 
+#Solves an issue with some system environments putting their MPI headers before
+#the headers CMake includes. Forces non-system MPI headers when incorrect headers
+#detected in include path.
+option(FENIX_SYSTEM_INC_FIX  "Attempts to force overriding any system MPI headers" ON)
 
-# Set empty string for shared linking (we use static library only at this moment)
-set(CMAKE_SHARED_LIBRARY_LINK_C_FLAGS)
 
-set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
-set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
+find_package(MPI REQUIRED)
 
-#set(CMAKE_BUILD_TYPE Release)
-set(CMAKE_BUILD_TYPE Debug)
-#set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -O0 -ggdb")
+#If we're using mpicc, we don't need to worry about the includes.
+if("${CMAKE_C_COMPILER}" MATCHES ".*/?mpicc")
+  set(FENIX_SYSTEM_INC_FIX OFF)
+endif()
 
-#ENABLE_TESTING
-set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${CMAKE_CURRENT_SOURCE_DIR})
-#include(testref/TestAgainstReference)
-
-configure_file(
-    ${CMAKE_CURRENT_SOURCE_DIR}/include/fenix-config.h.in
-    ${CMAKE_CURRENT_BINARY_DIR}/include/fenix-config.h @ONLY
-)
+if(${FENIX_SYSTEM_INC_FIX})
+  include(CheckIncludeFile)
+  set(CMAKE_REQUIRED_QUIET ON)
+  check_include_file("mpi.h" MPI_HEADER_CRASH)
+  set(CMAKE_REQUIRED_QUIET OFF)
 
+  if(${MPI_HEADER_CRASH})
+    message(WARNING "Detected system MPI headers, attempting to force use of ${MPI_C_INCLUDE_DIRS}. Set SYSTEM_INC_FIX=OFF to stop this behavior.")
 
-#Check for MPICC definition, if not try to find MPI
-if(NOT "a$ENV{MPICC}" STREQUAL "a")
-    #set(CMAKE_C_COMPILER ${MPI_C_COMPILER} CACHE STRING "The compiler CMake should use - often set to mpicc" FORCE)
-    set(MPI_C_COMPILER $ENV{MPICC})
-    set(CMAKE_C_COMPILER ${MPI_C_COMPILER})
-    
-    message("[fenix] MPICC has been passed: $ENV{MPICC}")
-else()
-    message("[fenix] MPICC was not passed, searching for MPI")
-    find_package(MPI REQUIRED)
-    if(${MPI_C_FOUND})
-        message("[fenix] Found MPICC: ${MPI_C_COMPILER}")
+    if(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.25")
+      set_target_properties(MPI::MPI_C PROPERTIES SYSTEM "FALSE")
```

Comment:
Does this actually work with >= 3.25? I thought that CMake would typically downgrade itself and not use newer features when the required CMake version was below the current.

Reply:
I gave it a test and it seems functional. Maybe it sets up backwards compatibility with the required version's features, but since this is a new feature it works out?